cosmos/listeners/task_instance_listener.py

```diff
@@ -0,0 +1,165 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Any
+
+from airflow.listeners import hookimpl
+
+if TYPE_CHECKING:
+    from airflow.models.taskinstance import TaskInstance
+
+from cosmos import telemetry
+from cosmos.constants import InvocationMode
+from cosmos.log import get_logger
+from cosmos.operators.base import AbstractDbtBase
+
+logger = get_logger(__name__)
+
+TASK_INSTANCE_EVENT = "task_instance"
+
+
+def _is_cosmos_task(task_instance: TaskInstance) -> bool:
+    """Return True if the task instance is powered by Cosmos operators."""
+
+    task = task_instance.task
+    module = _operator_module(task_instance)
+    return module.startswith("cosmos.") or isinstance(task, AbstractDbtBase)
+
+
+def _execution_mode_from_task(task_instance: TaskInstance) -> str | None:
+    """Extract Cosmos execution mode from the task's module path."""
+
+    module = _operator_module(task_instance)
+    parts = module.split(".")
+    if len(parts) >= 3 and parts[0] == "cosmos" and parts[1] == "operators":
+        return parts[2]
+    # TODO: When users subclass Cosmos operators in external modules, encode execution mode directly on the task
+    # so telemetry does not rely on module inspection.
+    return None
+
+
+def _operator_module(task_instance: TaskInstance) -> str:
+    """Return the module path for the operator backing the given task instance."""
+
+    return getattr(task_instance.task, "_task_module", None) or task_instance.task.__class__.__module__
```

Comment:
Which version of Airflow is related to each of these ways if retrieving the task module? It may be worth to document it
```suggestion
    return getattr(task_instance.task, "_task_module",  task_instance.task.__class__.__module__)
```